### PR TITLE
114 error with multiple available devices

### PR DESCRIPTION
**Pull request by Rob2U**

### CODE REVIEW

**Comment by Rob2U** [comment]
- registered attributes holding Tensors via `self.register_buffer` in all Bandits
- fixed Sherman-Morrison(-Woodbury) update formula
- added `eps` in LinearBandit to ensure 
- added `accelerator` to all notebooks*
- made `EpsilonGreedySelector` device agnostic
- removed unnecessary update logic (of LinearBandit) from NeuralLinear


* also tested -> only NeuralLinear could not be verified might be due to bugs in MPS (needs to be tested on cluster)

File: src/neural_bandits/bandits/linear_ts_bandit.py [hunk 1]
```diff
@@ -44,7 +44,7 @@ def _predict_action(
         ), "contextualized actions must have shape (batch_size, n_arms, n_features)"
         batch_size = contextualized_actions.shape[0]
 
-        theta_tilde = torch.distributions.MultivariateNormal(self.theta, self.precision_matrix).sample((batch_size,))  # type: ignore
+        theta_tilde = torch.distributions.MultivariateNormal(self.theta, precision_matrix=self.precision_matrix).sample((batch_size,))  # type: ignore
```
Comment (philippkolbe) [inline src/neural_bandits/bandits/linear_ts_bandit.py:47]: this looks like bad formatting?

File: src/neural_bandits/bandits/linear_bandit.py [hunk 2]
```diff
@@ -6,32 +6,40 @@
 
 
 class LinearBandit(AbstractBandit):
+    precision_matrix: torch.Tensor
+    b: torch.Tensor
+    theta: torch.Tensor
```
Comment (philippkolbe) [inline src/neural_bandits/bandits/linear_bandit.py:11]: is this type definition really correct - even though we are using register_buffer now?

File: tests/test_neural_linear.py [hunk 3]
```diff
@@ -106,15 +106,16 @@ def test_neural_linear_bandit_forward_small_sample_correct() -> None:
     output, _ = bandit(contextualized_actions)
     assert output.shape == (1, 2)
     # assert that the correct action is selected
-    assert torch.all(output == torch.tensor([[1, 0]]))
+    assert torch.all(output == torch.tensor([[0, 1]]))
 
     # now change the weights of the head to only regard the second feature (x2)
     bandit.theta = torch.tensor([0.0, 1.0])
+    bandit.precision_matrix = torch.tensor([[0.01, 0.0], [0.0, 0.01]])
```
Comment (philippkolbe) [inline tests/test_neural_linear.py:113]: if you are going to change the test (why did you change the test?) you will have to update the comment above

**Review by philippkolbe — APPROVED** [review]
lgtm

File: tests/test_neural_linear.py (hunk 3 above)
Comment (Rob2U) [reply tests/test_neural_linear.py:113]: I am actually confused why it passed previously. Right now, it should be correct. If `theta = torch.tensor([1.0, 0.0])` and:
```python
contextualized_actions = torch.tensor(
    [[[1.0, 0.0], [0.0, 1.0]]], dtype=torch.float32
)
```
Then the actions that should be chosen would be 0 and 1 and for the second case vice versa..

Comment (Rob2U) [reply tests/test_neural_linear.py:113]: The comments should still be correct

File: src/neural_bandits/bandits/linear_bandit.py (hunk 2 above)
Comment (Rob2U) [reply src/neural_bandits/bandits/linear_bandit.py:11]: I did not find a better way to do it. I cannot confirm, but `register_buffer` should in theory allow the same functionality as a `torch.Tensor`.

File: tests/test_neural_linear.py (hunk 3 above)
Comment (philippkolbe) [reply tests/test_neural_linear.py]: @Rob2U I fixed it. My calculation was correct (I forgot about the Linear encoder before but it basically does not matter).
The problem was that the precision was set to 0.01 which leads to a really high variance in the results. Somehow in this branch the seed must have changed, leading to an unfortunate theta_sample and the inverted result.

Comment (Rob2U) [reply tests/test_neural_linear.py]: Okay, I think I understood. I also made an error here `src/neural_bandits/bandits/linear_ts_bandit.py`.  `precision_matrix` is simply the wrong term. We want to use the inverse of the observed covariance matrix as the covariance for our LinearModel. This means that the `self.precision_matrix` refers to the precision matrix of the observed contexts but not the distribution. This the call to `MultivariateNormal` in `linear_ts_bandit.py` is wrong leading to a very high variance, which in turn leads to the test failure.

Comment (Rob2U) [reply tests/test_neural_linear.py]: Will fix this in the next commit
